src/core/inductive.ml

```diff
@@ -0,0 +1,263 @@
+(** Generating of induction principles. *)
+
+open Timed
+open Pos
+open Console
+open Terms
+open Print
+open Syntax
+
+(** Logging function for generating of inductive principle. *)
+let log_ind = new_logger 'g' "indu" "generating induction principle"
+let log_ind = log_ind.logger
+
+(** Builtin configuration for induction. *)
+type config =
+  { symb_Prop : sym (** : TYPE. Type of propositions. *)
+  ; symb_prf  : sym (** : Prop → TYPE.
+                        Interpretation of propositions as types. *) }
+
+(** [get_config ss pos] build the configuration using [ss]. *)
+let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
+  let builtin = Builtin.get ss pos in
+  { symb_Prop = builtin "Prop"
+  ; symb_prf  = builtin "P" }
+
+(** [principle ss pos sind scons_list] returns an induction principle which
+    is created thanks to the symbol of the inductive type [sind] (and its
+    position [pos]), its constructors [scons_list] and the signature [ss]. *)
+let principle : Sig_state.t -> popt -> sym -> sym list -> term =
+  fun ss pos sind scons_list ->
+  let c = get_config ss pos in
+  let ind = _Symb sind in
+  let prop = _Symb c.symb_Prop in
+  let p = Bindlib.new_var mkfree "p" in
+  let prf_of_p t = _Appl (_Symb c.symb_prf) (_Appl (_Vari p) t) in
+  let app = List.fold_left _Appl in
+
+  (* [case_of scons] creates a clause according to a constructor [scons]. *)
+  let case_of : sym -> tbox = fun scons ->
+    let rec case : tbox list -> term-> tbox = fun xs a ->
+      match Basics.get_args a with
+      | (Symb(s), l) ->
+          if s == sind then
+            let l = List.map lift l in
+            match List.rev l with
```

Comment:
Why do you distinguish these two cases?

Reply:
And why do you reverse the list of arguments?

Reply:
Idem below.

Reply:
In fact, it seems to me that, wrt the previous commit, you just need to redefine prf_of_p to take arguments into account:
```
(* it compiles *)
  let app : tbox -> tbox list -> tbox = List.fold_left _Appl in
  let fapp : sym -> tbox list -> tbox = fun f ts -> app (_Symb f) ts in
  let prf_of_p ts t = fapp c.symb_prf [_Appl (app (_Vari p) ts) t] in

  (* [case_of scons] creates a clause according to a constructor [scons]. *)
  let case_of : sym -> tbox = fun scons ->
    let rec case : tvar list -> term-> tbox = fun xs a ->
      match Basics.get_args a with
      | (Symb(s), ts) ->
          if s == sind then
            prf_of_p (List.map lift ts) (fapp scons (List.rev_map _Vari xs))
          else fatal pos "%a is not a constructor of %a"
                 pp_symbol scons pp_symbol sind
      | (Prod(a,b), _) ->
          let (x,b) =
            if Bindlib.binder_occur b then
              Bindlib.unbind b
            else
              let x = Bindlib.new_var mkfree "x" in
              (x, Bindlib.subst b (Vari x))
          in
          let b = case (x::xs) b in
          begin
            match Basics.get_args a with
            | (Symb(s), ts) ->
                let b =
                  if s == sind then
                    _Impl (prf_of_p (List.map lift ts) (_Vari x)) b
                  else b
                in
                if !log_enabled then
                  log_ind "The current type : %a" Print.pp_term a;
                let tmp = _Prod (Bindlib.box a) (Bindlib.bind_var x b) in
                if !log_enabled then
                  log_ind "The clause of the constructor [%a] is %a"
                    Print.pp_term (Symb(scons))
                    Print.pp_term (Bindlib.unbox tmp);
                tmp
            | _ -> fatal pos "The type of %a is not supported"
                     pp_symbol scons
          end
      | _ ->
          fatal pos "The type of %a is not supported"
            pp_symbol scons
    in
```

Reply:
It's because I need to create an application : \pi ( p l1 ... ln t ) where l1 ... ln are the elements of the list l. So you need to distinguish if I is empty or not.

Reply:
Well, as shown in my code above, this is not necessary. Thanks to fix it.

Reply:
Yes you right, it's OK now.